source/common/network/io_socket_error_impl.cc

```diff
@@ -12,33 +12,26 @@ Api::IoError::IoErrorCode IoSocketError::getErrorCode() const { return error_cod
 
 std::string IoSocketError::getErrorDetails() const { return errorDetails(errno_); }
 
-IoSocketError* IoSocketError::getIoSocketInvalidAddressInstance() {
-  static auto* instance =
-      new IoSocketError(SOCKET_ERROR_NOT_SUP, Api::IoError::IoErrorCode::NoSupport);
-  return instance;
+Api::IoErrorPtr IoSocketError::getIoSocketInvalidAddressError() {
+  return Api::IoError::wrap(
+      new IoSocketError(SOCKET_ERROR_NOT_SUP, Api::IoError::IoErrorCode::NoSupport));
```

Comment:
Why not return a singleton for this error anymore?

Reply:
I think the original of this was just unconsidered copypaste from `Eagain`. It's not particularly justified for this one to use a singleton since it's not an expected, recurring error, so it doesn't need the performance boost. IMO we'd either want to make *all* of the expected errors singletons, or it should only be Eagain, as the one that has comments justifying it.

Reply:
A little prefer to using singleton. since if there is nothing bad, maybe we can just return singleton, then in the future we won't care about it anymore. But yes, I guess the pattern won't change too much if this error won't be used too much in the future also

Reply:
One issue with the singleton is that it's just generally poor hygiene as essentially a global heap variable that is never deleted - it's bad for testing, for example, in that the behavior in the first test is not actually the same as the behavior in the second test, because cleanup doesn't reset things to original state. And it's *probably* a cause of new linter errors about things not getting deleted (though using 'new' with custom deleters at all is probably enough to provoke that - in the singleton case the linter is *correct* to complain about something never getting deleted.)

That's the main reason I opted for not propagating the singleton pattern into the contexts where it's not actually beneficial - the less we make that special-case exception the better.